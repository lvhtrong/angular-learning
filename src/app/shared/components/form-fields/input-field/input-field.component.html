<div class="form-group">
  <label [for]="id">{{ label }}</label>
  <input
    [id]="id"
    [type]="type"
    [attr.data-testid]="testId"
    class="form-control"
    [class.is-invalid]="error"
    [value]="value"
    (input)="onTextChange($event.target.value)"
  />
  <div class="invalid-feedback">
    {{ error }}
  </div>
</div>
