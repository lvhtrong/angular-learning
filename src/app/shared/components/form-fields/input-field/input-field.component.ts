import { Component, OnInit, Output, Input } from '@angular/core';

@Component({
  selector: 'app-f-input-field',
  templateUrl: './input-field.component.html',
  styleUrls: ['./input-field.component.scss'],
})
export class InputFieldComponent implements OnInit {
  @Input() id = 'input-field';
  @Input() label = '';
  @Input() value = '';
  @Input() type = 'text';
  @Input() error: string;
  @Output() textChange: (value: string) => {};

  constructor() {}

  ngOnInit(): void {}

  onTextChange(value: string) {
    this.textChange(value);
  }
}
