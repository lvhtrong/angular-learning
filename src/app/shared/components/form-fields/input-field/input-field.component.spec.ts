import { createHostFactory, Spectator, byLabel } from '@ngneat/spectator/jest';

import { InputFieldComponent } from './input-field.component';

describe('InputFieldComponent', () => {
  const createComponent = createHostFactory({
    component: InputFieldComponent,
  });

  const setup = (props?: any): Spectator<InputFieldComponent> => {
    return createComponent('<app-f-input-field></app-f-input-field>', {
      props,
    });
  };

  it('should create', () => {
    const spectator = setup();
    expect(spectator).toBeTruthy();
  });

  it('should render with inputs', () => {
    const spectator = setup();
    spectator.setInput('id', 'testId');
    spectator.setInput('label', 'sample label');
    spectator.setInput('value', 'sample value');
    spectator.setInput('type', 'email');
    spectator.setInput('error', 'sample error');

    expect(spectator.element).toMatchSnapshot();
  });

  it('should invoke textChange event when typing', () => {
    const expectedValue = 'expectedValue';

    const textChange = jest.fn();
    const spectator = setup({
      textChange,
    });

    const fieldLabel = 'field-label';
    spectator.setInput('label', fieldLabel);

    const inputElement = spectator.query(byLabel(fieldLabel));
    spectator.typeInElement(expectedValue, inputElement);

    expect(textChange).toHaveBeenCalledTimes(1);
    expect(textChange).toHaveBeenCalledWith(expectedValue);
  });

  describe('error', () => {
    const label = 'Label';

    const setupError = () => {
      const spectator = setup();
      spectator.setInput('label', label);

      return spectator;
    };

    it('should have class "is-invalid"', () => {
      const spectator = setupError();
      spectator.setInput('error', 'error');

      const inputElement = spectator.query(byLabel(label));
      expect(inputElement).toHaveClass('is-invalid');
    });

    it('should not have class "is-invalid"', () => {
      const spectator = setupError();

      const inputElement = spectator.query(byLabel(label));
      expect(inputElement).not.toHaveClass('is-invalid');
    });
  });
});
