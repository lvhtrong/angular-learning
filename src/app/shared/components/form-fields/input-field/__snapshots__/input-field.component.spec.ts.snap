// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFieldComponent should render with inputs 1`] = `
<app-f-input-field>
  <div
    class="form-group"
  >
    <label
      for="testId"
    >
      sample label
    </label>
    <input
      class="form-control is-invalid"
      id="testId"
      type="email"
    />
    <div
      class="invalid-feedback"
    >
       sample error 
    </div>
  </div>
</app-f-input-field>
`;
